File: lib/rest/yampi/services/yampi_checkout_service.dart
import 'package:sertton/constants/env.dart';
import 'package:sertton/core/checkout/dtos/installment_dto.dart';
import 'package:sertton/core/checkout/dtos/order_dto.dart';
import 'package:sertton/core/checkout/dtos/payment_dto.dart';
import 'package:sertton/core/checkout/interfaces/checkout_service.dart';
import 'package:sertton/core/global/responses/rest_response.dart';
import 'package:sertton/rest/yampi/mappers/yampi_order_mapper.dart';
import 'package:sertton/rest/yampi/mappers/yampi_payment_mapper.dart';
import 'package:sertton/rest/yampi/mappers/yampi_installment_mapper.dart';
import 'package:sertton/rest/yampi/services/yampi_service.dart';

class YampiCheckoutService extends YampiService implements CheckoutService {
  YampiCheckoutService(super.restClient, super.envDriver);

  @override
  Future<RestResponse<String>> fetchCheckoutLink(
    List<String> skuTokens,
    List<int> quantities,
  ) async {
    if (skuTokens.isEmpty || quantities.isEmpty) {
      throw ArgumentError('skuTokens and quantities cannot be empty');
    }

    if (skuTokens.length != quantities.length) {
      throw ArgumentError('skuTokens and quantities must have the same length');
    }

    final items = <String>[];
    for (var index = 0; index < skuTokens.length; index++) {
      items.add('${skuTokens[index]}:${quantities[index]}');
    }

    final itemsParam = items.join(',');

    final link = Uri.parse(
      '${super.envDriver.get(Env.yampiPurchaseUrl)}/$itemsParam',
    );

    return RestResponse(body: link.toString());
  }

  @override
  Future<RestResponse<List<OrderDto>>> fetchOrdersByCustomer(
    String customerDocument,
  ) async {
    final response = await super.restClient.get(
      '/orders',
      queryParams: {'q': customerDocument},
    );

    return response.mapBody((json) => YampiOrderMapper.toDtoList(json));
  }

  @override
  Future<RestResponse<List<PaymentDto>>> fetchPayments() async {
    final response = await super.restClient.get('/checkout/payments');

    return response.mapBody((json) => YampiPaymentMapper.toDtoList(json));
  }

  @override
  Future<RestResponse<List<InstallmentDto>>> fetchInstallments(
    String paymentId,
    String productId,
    double productPrice,
  ) async {
    final response = await super.restClient.get(
      '/public/catalog/products/$productId/installments',
      queryParams: {'amount': productPrice},
    );

    return response.mapBody((json) => YampiInstallmentMapper.toDtoList(json));
  }
}

File: src/rest/yampi/services/yampi-checkout-service.ts
import type { ICheckoutService } from "@/core/checkout/interfaces/checkout-service"
import { ValidationError } from "@/core/shared/errors/app-error"
import type { RestClient } from "@/core/shared/interfaces/rest-client"
import { RestResponse } from "@/core/shared/responses/rest-response"
import { YampiInstallmentMapper, YampiOrderMapper, YampiPaymentMapper } from "@/rest/yampi/mappers"
import type {
  YampiInstallmentsResponse,
  YampiOrder,
  YampiPayment,
  YampiResponse,
} from "@/rest/yampi/types"

export interface YampiCheckoutDependencies {
  readonly purchaseUrl: string
  readonly restClient: RestClient
}

export const YampiCheckoutService = ({
  purchaseUrl,
  restClient,
}: YampiCheckoutDependencies): ICheckoutService => {
  const installmentMapper = YampiInstallmentMapper()
  const orderMapper = YampiOrderMapper()
  const paymentMapper = YampiPaymentMapper()

  return {
    async fetchCheckoutLink(skuTokens, quantities) {
      if (
        skuTokens.length === 0 ||
        quantities.length === 0 ||
        skuTokens.length !== quantities.length
      )
        throw new ValidationError("Invalid checkout items")
      return new RestResponse({
        body: `${purchaseUrl.replace(/\/$/, "")}/${skuTokens.map((token, index) => `${token}:${quantities[index]}`).join(",")}`,
      })
    },

    async fetchInstallments(paymentId, productId, productPrice) {
      void paymentId
      restClient.clearQueryParams()
      restClient.setQueryParam("amount", productPrice)
      const response = await restClient.get<YampiInstallmentsResponse>(
        `/public/catalog/products/${productId}/installments`,
      )
      return response.mapBody((body) => body.installments.map(installmentMapper.toDomain))
    },

    async fetchOrdersByCustomer(customerDocument) {
      const response = await restClient.get<YampiResponse<YampiOrder>>("/orders", {
        query: {
          customer_document: customerDocument,
          include: "items,shipping_address,status,transactions",
        },
      })
      return response.mapBody((body) => body.data.map(orderMapper.toDomain))
    },

    async fetchPayments() {
      const response = await restClient.get<YampiResponse<YampiPayment>>("/checkout/payments")
      return response.mapBody((body) => body.data.map(paymentMapper.toDomain))
    },
  }
}
